elastic/kibana
# [Watcher] More robust handling of license at setup

## Summary

Fix after https://github.com/elastic/kibana/pull/54752 was merged. Watcher, at setup time, subscribes to a `licensing.license$` stream and only registers after it has received the first message from that stream. The problem is the setup function is running synchronously and this was causing some flakiness in ordering. One issue I saw was (when navigating to Watcher):

<img width="633" alt="Screenshot 2020-01-24 at 12 25 11" src="https://user-images.githubusercontent.com/8155004/73068477-a3673080-3eab-11ea-81b3-df079e96df86.png">

This would redirect to home page as if the app has been disabled (even though it was not).

Additionally there was some jankiness with when the app link would be shown.

## Changes

- ~In `plugin.ts` we now have an `async` setup method which will await the first license from `licensing.license$`. Only after it has a valid license will it process app registration.~ `setup` should remain synchronous https://github.com/elastic/kibana/pull/45796#discussion_r327830578. We now do all registration synchronously and then respond to `license$` updates.
- In `app.tsx` we receive an initial license value and an observable so that we can react to license updates and handle license changes in a more graceful way.

## How to test

1. Navigate to watcher (w/ min. license)
2. Run `curl -XDELETE -v http://elastic:changeme@localhost:9200/_license && curl -XPOST -v http://elastic:changeme@localhost:9200/_license/start_basic\?acknowledge\=true`
3. Wait roughly 1 minute and the watcher UI should update with a spinner (it's trying to refresh but can't), then the license update should change the UI to callout explaining the issue.
4. Refresh and watcher should not be in ES section

If you don't have the required min. license in place for watcher, the link should not be there in the ES management section.

## Additional Notes

- You can run into this bug https://github.com/elastic/kibana/issues/55170 while testing. Just manually update the address bar to management section.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- ~This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)~
- ~Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)~
- ~[Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials~
- ~[Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios~
- ~This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)~

### For maintainers

- ~This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)~
